x/gamm/twap/README.md

```diff
@@ -1,9 +1,137 @@
+// the state machine will interpolate the accumulator values for those times
+// from the latest Twap accumulation record prior to the provided time.
+//
+// startTime must be within 48 hours of ctx.BlockTime(), if you need older TWAPs,
+// you will have to maintain the accumulator yourself.
+//
+// This function will error if:
+// * startTime > endTime
+// * endTime in the future
+// * startTime older than 48 hours OR pool creation
+// * pool with id poolId does not exist, or does not contain quoteAssetDenom, baseAssetDenom
+//
+// N.B. If there is a notable use case, the state machine could maintain more historical records, e.g. at one per hour.
+func (k Keeper) GetArithmeticTwap(ctx sdk.Context,
+	poolId uint64,
+	baseAssetDenom string, quoteAssetDenom string,
+	startTime time.Time, endTime time.Time) (sdk.Dec, error) { ... }
+```
+
+There are convenience methods for `GetArithmeticTwapToNow` which sets `endTime = ctx.BlockTime()`, and has minor gas reduction.
+For users who need TWAPs outside the 48 hours stored in the state machine, you can get the latest accumulation store record from `GetBeginBlockAccumulatorRecord`.
+
+## Code layout
+
+**api.go** is the main file you should look at as a user of this module.
+
+**logic.go** is the main file you should look at for how the TWAP implementation works.
+
+- types/* - Implement TwapRecord, GenesisState. Define AMM interface, and methods to format keys.
+- api.go - Public API, that other users / modules can/should depend on
+- hook_listener.go - Defines hooks & calls to logic.go, for triggering actions on 
+- keeper.go - generic SDK boilerplate (defining a wrapper for store keys + params)
+- logic.go - Implements all TWAP module 'logic'. (Arithmetic, defining what to get/set where, etc.)
+- module.go - SDK AppModule interface implementation.
+- store.go - Managing logic for getting and setting things to underlying stores
+
+## Store layout
+
+We maintain twap accumulation records for every AMM pool on Osmosis. 
+
+Because Osmosis supports multi-asset pools, a complicating factor is that we have to store a record for every asset pair in the pool.
+For every pool, at a given point in time, we make one twap record entry per unique pair of denoms in the pool. If a pool has `k` denoms, the number of unique pairs is `k * (k - 1) / 2`.
+
+Each twap record stores [(source)](https://github.com/osmosis-labs/osmosis/tree/main/proto/osmosis/gamm/twap):
+* last spot price of base asset A in terms of quote asset B
+* last spot price of base asset B in terms of quote asset A
+* Accumulation value of base asset A in terms of quote asset B
+* Accumulation value of base asset B in terms of quote asset A
+
+All TWAP records are indexed in state by the time of write.
+
+A new TWAP record is created in two situations:
+* When a pool is created
+* In the `EndBlock`, if the block contains any potentially price changing event for the pool. (Swap, LP, Exit)
```

Comment:
Are (Swap, LP, Exit) exhaustive? Suggest specifying if this is an example or an exhaustive list